spark/src/test/java/org/apache/iceberg/TestTableSerialization.java

```diff
@@ -0,0 +1,221 @@
+
+      ByteArrayOutputStream bytes = new ByteArrayOutputStream();
+      try (ObjectOutputStream out = new ObjectOutputStream(bytes)) {
+        out.writeObject(metadataTable);
+      }
+
+      try (ObjectInputStream in = new ObjectInputStream(new ByteArrayInputStream(bytes.toByteArray()))) {
+        Object obj = in.readObject();
+        Assert.assertTrue("Should be a Table", obj instanceof Table);
+        checkTable(metadataTable, (Table) obj);
+      }
+    }
+  }
+
+  @Test
+  public void testSerializedMetadataTableKryoSerialization() throws IOException {
+    for (MetadataTableType type : MetadataTableType.values()) {
+      File data = temp.newFile();
+      Assert.assertTrue(data.delete());
+
+      Kryo kryo = new KryoSerializer(new SparkConf()).newKryo();
+
+      TableOperations ops = ((HasTableOperations) table).operations();
+      Table metadataTable = MetadataTableUtils.createMetadataTableInstance(ops, table.name(), "meta", type);
+      Table serializedMetadataTable = SparkUtil.toSerializedTable(metadataTable);
+      checkTable(metadataTable, serializedMetadataTable);
+
+      try (Output out = new Output(new FileOutputStream(data))) {
+        kryo.writeClassAndObject(out, serializedMetadataTable);
+      }
+
+      try (Input in = new Input(new FileInputStream(data))) {
+        Object obj = kryo.readClassAndObject(in);
+        Assert.assertTrue("Should be a Table", obj instanceof Table);
+        checkTable(metadataTable, (Table) obj);
+      }
+    }
+  }
+
+  @Test
+  public void testSerializedMetadataTableJavaSerialization() throws IOException, ClassNotFoundException {
+    for (MetadataTableType type : MetadataTableType.values()) {
+      File data = temp.newFile();
+      Assert.assertTrue(data.delete());
+
+      TableOperations ops = ((HasTableOperations) table).operations();
+      Table metadataTable = MetadataTableUtils.createMetadataTableInstance(ops, table.name(), "meta", type);
+      Table serializedMetadataTable = SparkUtil.toSerializedTable(metadataTable);
+      checkTable(metadataTable, serializedMetadataTable);
+
+      ByteArrayOutputStream bytes = new ByteArrayOutputStream();
+      try (ObjectOutputStream out = new ObjectOutputStream(bytes)) {
+        out.writeObject(serializedMetadataTable);
+      }
+
+      try (ObjectInputStream in = new ObjectInputStream(new ByteArrayInputStream(bytes.toByteArray()))) {
+        Object obj = in.readObject();
+        Assert.assertTrue("Should be a Table", obj instanceof Table);
+        checkTable(metadataTable, (Table) obj);
+      }
+    }
+  }
+
+  private void checkTable(Table expected, Table actual) {
+    Assert.assertEquals("Name must match", expected.name(), actual.name());
+    Assert.assertEquals("Location must match", expected.location(), actual.location());
+    Assert.assertEquals("Props must match", expected.properties(), actual.properties());
+    Assert.assertEquals("Schema must match", expected.schema().asStruct(), actual.schema().asStruct());
```

Comment:
This will now fail, unfortunately. The way we compare struct fields is broken with Kryo serialization. We rely on the `PrimitiveHolder` proxy but Kryo ignores that.

Reply:
This has been fixed. Resolving.